lisp/ess-inf.el

```diff
@@ -1293,43 +1293,44 @@ wrapping the code into:
-                (setq buffer-read-only nil)
-                (erase-buffer)
-                (set-marker (process-mark proc) (point-min))
-                (inferior-ess-mark-as-busy proc)
-                (process-send-string proc rich-cmd)
-                ;; Need time for ess-create-object-name-db on PC
-                (if no-prompt-check
-                    (sleep-for 0.02)   ; 0.1 is noticeable!
-                  (unless (ess-wait-for-process proc nil wait force-redisplay timeout)
-                    (error "Timeout during background ESS command `%s'"
-                           (ess--strip-final-newlines cmd)))
-                  ;; Remove prompt. If output is cat(..)ed without a
-                  ;; final newline, this deletes the last line of output.
+            (unwind-protect
+                (progn
+                  (when use-sentinel
+                    (process-put proc 'ess-output-sentinel sentinel))
+                  (set-process-buffer proc out-buffer)
+                  (set-process-filter proc 'inferior-ess-ordinary-filter)
+                  (with-current-buffer out-buffer
+                    (ess-setq-vars-local proc-alist)
+                    (setq buffer-read-only nil)
+                    (erase-buffer)
+                    (set-marker (process-mark proc) (point-min))
+                    (inferior-ess-mark-as-busy proc)
+                    (process-send-string proc rich-cmd)
+                    ;; Need time for ess-create-object-name-db on PC
+                    (if no-prompt-check
+                        (sleep-for 0.02) ; 0.1 is noticeable!
+                      (unless (ess-wait-for-process proc nil wait force-redisplay timeout)
+                        (error "Timeout during background ESS command `%s'"
+                               (ess--strip-final-newlines cmd)))
+                      ;; Remove prompt. If output is cat(..)ed without a
+                      ;; final newline, this deletes the last line of output.
+                      (goto-char (point-max))
+                      (delete-region (point-at-bol) (point-max))))
+                  (setq early-exit nil))
+              (when early-exit
+                (with-current-buffer out-buffer
+                  (goto-char (point-min))
+                  (when (and use-sentinel
+                             (not (re-search-forward
+                                   (inferior-ess--sentinel-start-re sentinel)
+                                   nil t)))
+                    ;; CMD probably failed to parse if the start sentinel
+                    ;; can't be found in the output. Disable the sentinel
+                    ;; before interrupt to avoid a freeze.
+                    (process-put proc 'ess-output-sentinel nil))
                   (goto-char (point-max))
-                  (delete-region (point-at-bol) (point-max))))
-              (setq early-exit nil))
+                  (ess-interrupt))))
```

Comment:
We use `ess-interrupt` so we block until prompt shows up. If we used `interrupt-process` we wouldn't block and output would spill in the process buffer.

Reply:
Timeouts are not a code smell to me, they are essential to concurrency with external entities. I really think we need to block with a timeout. The user may have already typed `C-g` and if ESS doesn't respond to the first interrupt consistently, sometimes requiring multiple interrupts, this deteriorates the user sense of being in control of the process.

I think using `ess-interrupt` rather than `interrupt-process` is cleaner from the standpoint of code reuse consolidating internals.

I don't know exactly what you're referring to with the `"\n"` but if it's tangential I'd prefer to focus on the main points.

Reply:
It's not tangential. Look at the `ess-interrupt` code. 

> I think using ess-interrupt rather than interrupt-process is cleaner from the standpoint of code reuse and consolidating internals.

Sometimes using lower level utilities in critical code is cleaner and beter conveys the meaning. `ess-interrupt` in this case does a bunch of things that are intended for interactive use. There is no place for things like `(inferior-ess-force)`, `(ess-get-process)` `(timeout 1)` `(process-send-string nil "\n")` `(goto-char (process-mark proc))` in `ess-command` IMO.

Reply:
Are you sure that `"\n"` is not needed on Windows terminals? Since we are blocking for prompt it seems risky not to send it? While I see your point, I think my point holds as well, considering the uncertainty about cross-platform behaviour.

Reply:
I have refactored `ess-interrupt` into interactive and non-interactive functions. `ess-command` now uses the non-interactive one. I kept the newline thing in the core routine because of the uncertainty regarding its purpose.